tests/integration/retries.py

```diff
@@ -0,0 +1,214 @@
+    if dilation != 1.0:
+        logger.info(f"Non-default timeout dilation: all timeouts will be scaled by a factor of {dilation}.")
+    return dilation
+
+
+P = ParamSpec("P")
+T = TypeVar("T")
+
+
+class _Retrier:
+    __slots__ = ("_on", "_is_retryable", "_timeout", "_clock", "_min_attempts", "_max_attempts")
+
+    def __init__(
+        self,
+        *,
+        on: Sequence[type[BaseException]] | None,
+        is_retryable: Callable[[BaseException], str | None] | None,
+        timeout: dt.timedelta | None,
+        clock: Clock,
+        timeout_dilation: float,
+        min_attempts: int,
+        max_attempts: int | None = None,
+    ) -> None:
+        has_allowlist = on is not None
+        has_callback = is_retryable is not None
+        if has_callback == has_allowlist:
+            raise SyntaxError('either on=[Exception] or callback=lambda x: .. is required')
+        if timeout is None and max_attempts is None:
+            raise SyntaxError('either timeout=... or max_attempts=... is required')
+        if max_attempts is not None and max_attempts < min_attempts:
+            msg = f'max_attempts=x must be equal or higher to min_attempts ({min_attempts})'
+            raise SyntaxError(msg)
+
+        self._on = on
+        self._is_retryable = is_retryable
+
+        dilated_timeout = timeout * timeout_dilation if timeout is not None else None
+        self._timeout = dilated_timeout
+        self._clock = clock
+        self._min_attempts = min_attempts
+        self._max_attempts = max_attempts
+
+    def _can_retry(self, err: BaseException) -> str | None:
+        if self._on is not None:
+            return self._subclass_checker(err)
+        if self._is_retryable is not None:
+            return self._is_retryable(err)
+        # Unreachable
+        return None
+
+    def _subclass_checker(self, err: BaseException) -> str | None:
+        assert self._on is not None
+        for candidate in self._on:
+            if isinstance(err, candidate):
+                return f'{type(err).__name__} is allowed to retry'
+        return None
+
+    def _deadline(self) -> float | None:
+        timeout = self._timeout
+        return self._clock.time() + timeout.total_seconds() if timeout is not None else None
+
+    def _check_attempt_duration(self, func: Callable[P, T], attempt_duration: dt.timedelta) -> None:
+        timeout = self._timeout
+        if timeout is not None:
+            if timeout <= attempt_duration:
+                logger.error(
+                    f"Retryable call ({func.__name__}) attempt took {attempt_duration}, longer than retry timeout ({timeout}). Adjust timeout accordingly!"
+                )
+            elif (timeout * 0.80) <= attempt_duration:
+                logger.warning(
+                    f"Retryable call ({func.__name__}) attempt took {attempt_duration}, within 20% of the specified timeout ({timeout}); please adjust timeout accordingly."
```

Comment:
```suggestion
                    f"Retryable call ({func.__name__}) attempt took {attempt_duration}, within 80% of the specified timeout ({timeout}); please adjust timeout accordingly."
```

Reply:
Hmm. That's funny, because 20% is what I intended here and I'm pretty sure it's the correct usage in conjunction with _within_. But when I saw this suggestion I definitely had to think about it, and I think it's ambiguous what the 'within 20%' refers to. I've adjusted the wording to (hopefully!) make it clearer.